File: libs/contracts/report.py
"""Báo cáo lĩnh vực & báo cáo tổng hợp — xương sống đánh nguồn kiểu NotebookLM.

Bất biến bắt buộc (chống bịa nguồn):
- Trong `DomainReport`: mọi `citation_id` mà một `Claim` tham chiếu PHẢI khớp
  `Citation.id` có thật trong `citations` của chính báo cáo đó.
- Ở `ComposedReport`: mọi `citation_id` của luận điểm tổng hợp PHẢI truy được về
  một `Citation` gốc do specialist tạo (verify ở tầng Compose). Citation không
  resolve được bị loại và luận điểm đánh dấu `unsupported`.
"""

from __future__ import annotations

from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator

from libs.contracts._enum import StrEnum
from libs.contracts.citation import Citation


class Domain(StrEnum):
    CREDIT = "CREDIT"      # tài chính (REAL, +SQL tool)
    LEGAL = "LEGAL"        # pháp lý / chính sách (MOCK)
    DOCUMENT = "DOCUMENT"  # hồ sơ / OCR (MOCK)


class ReportStatus(StrEnum):
    OK = "OK"            # specialist chạy trọn vẹn
    PARTIAL = "PARTIAL"  # thiếu dữ liệu / một phần lỗi, vẫn có kết quả dùng được
    FAILED = "FAILED"    # specialist lỗi hẳn (orchestrator vẫn compose phần còn lại)


class Claim(BaseModel):
    """Một luận điểm của specialist; mỗi luận điểm phải có nguồn (citation_ids)."""

    model_config = ConfigDict(extra="forbid")

    text: str
    citation_ids: tuple[str, ...] = ()
    # đánh dấu luận điểm suy luận không trực tiếp từ nguồn (compose sẽ hạ trọng số)
    is_inference: bool = False


class DomainReport(BaseModel):
    """Đầu ra chuẩn của MỖI specialist (Credit/Legal/Document).

    MOCK và REAL trả cùng schema này → cắm microservice thật không đổi Compose.
    """

    model_config = ConfigDict(extra="forbid")

    domain: Domain
    agent: str
    status: ReportStatus = ReportStatus.OK
    summary: str = ""
    claims: tuple[Claim, ...] = ()
    citations: tuple[Citation, ...] = ()
    confidence: float = Field(default=0.7, ge=0.0, le=1.0)
    error: str | None = None

    @model_validator(mode="after")
    def _claims_cite_real_citations(self) -> "DomainReport":
        known = {c.id for c in self.citations}
        for claim in self.claims:
            for cid in claim.citation_ids:
                if cid not in known:
                    raise ValueError(
                        f"[{self.domain}] claim trích citation không tồn tại: {cid!r}"
                    )
        # Citation kiểu DOCUMENT bắt buộc có bbox để highlight được.
        for c in self.citations:
            if c.source_type == "DOCUMENT" and c.bbox is None:
                raise ValueError(f"citation DOCUMENT {c.id!r} thiếu bbox")
        return self


class ComposedClaim(BaseModel):
    """Luận điểm trong báo cáo tổng hợp; citation_ids trỏ về Citation gốc specialist."""

    model_config = ConfigDict(extra="forbid")

    text: str
    citation_ids: tuple[str, ...] = ()
    supported: bool = True  # False nếu verify không resolve được nguồn


class ComposedSection(BaseModel):
    model_config = ConfigDict(extra="forbid")

    title: str
    domain: Domain | None = None
    claims: tuple[ComposedClaim, ...] = ()


class ComposedReport(BaseModel):
    """Báo cáo tổng hợp cho người duyệt — đã verify chống bịa nguồn."""

    model_config = ConfigDict(extra="forbid")

    correlation_id: UUID
    question: str
    recommendation: str = ""
    sections: tuple[ComposedSection, ...] = ()
    # Citation gốc gộp từ mọi specialist (nguồn chân lý để UI resolve → highlight).
    citations: tuple[Citation, ...] = ()
    # Luận điểm bị loại nguồn ở bước verify (minh bạch, không giấu).
    unsupported_claims: tuple[str, ...] = ()
    domain_status: dict[str, ReportStatus] = Field(default_factory=dict)
    eval_rounds: int = 0
    warnings: tuple[str, ...] = ()

    @model_validator(mode="after")
    def _composed_claims_resolve(self) -> "ComposedReport":
        known = {c.id for c in self.citations}
        for section in self.sections:
            for claim in section.claims:
                for cid in claim.citation_ids:
                    if cid not in known:
                        raise ValueError(
                            f"composed claim trích citation không có trong nguồn gộp: {cid!r}"
                        )
                if claim.citation_ids and not claim.supported:
                    raise ValueError("claim có citation_ids thì không được supported=False")
        return self


def collect_citations(reports: list[DomainReport]) -> tuple[Citation, ...]:
    """Gộp toàn bộ Citation gốc từ các specialist (giữ thứ tự, khử trùng id)."""
    seen: set[str] = set()
    out: list[Citation] = []
    for report in reports:
        for c in report.citations:
            if c.id not in seen:
                seen.add(c.id)
                out.append(c)
    return tuple(out)

File: frontend/src/types/report.ts
import type { SourceLocator } from './source';
import type { DocumentRegion } from '@/features/chat/constants/demoDocuments';

/**
 * Báo cáo của chuyên gia số — PHẢN CHIẾU ĐÚNG hợp đồng backend
 * (`libs/contracts/report.py` + `citation.py`).
 *
 * Giữ đúng hình dạng này là một trong ba rủi ro đã khoá của dự án: "mock = hợp
 * đồng thật". Khi cắm Engine thật vào, giao diện không phải sửa.
 *
 * Bất biến bắt buộc (chống bịa nguồn):
 * - Mọi `citationIds` của một luận điểm PHẢI khớp `ReportCitation.id` có thật
 *   trong `citations` của chính báo cáo đó.
 * - Citation kiểu DOCUMENT BẮT BUỘC có `locator` để giao diện highlight được.
 */

export type ReportDomain = 'CREDIT' | 'LEGAL' | 'DOCUMENT';

export const REPORT_DOMAIN_LABEL: Record<ReportDomain, string> = {
  CREDIT: 'Tài chính',
  LEGAL: 'Pháp lý & Tuân thủ',
  DOCUMENT: 'Hồ sơ & Chứng từ',
};

export type ReportStatus = 'OK' | 'PARTIAL' | 'FAILED';

export const REPORT_STATUS_LABEL: Record<ReportStatus, string> = {
  OK: 'Hoàn tất',
  PARTIAL: 'Thiếu dữ liệu một phần',
  FAILED: 'Không chạy được',
};

/** Loại nguồn — quyết định cách giao diện mở dẫn chứng. */
export type CitationSourceType = 'DOCUMENT' | 'POLICY' | 'RECORD' | 'SQL';

export const CITATION_SOURCE_LABEL: Record<CitationSourceType, string> = {
  DOCUMENT: 'Tài liệu gốc',
  POLICY: 'Điều khoản chính sách',
  RECORD: 'Bản ghi hệ thống',
  SQL: 'Truy vấn dữ liệu',
};

export interface ReportCitation {
  id: string;
  sourceType: CitationSourceType;
  /** Đoạn nguyên văn được trích — hiện nguyên si, không diễn giải lại. */
  quote: string;
  /** Có mặt khi sourceType = DOCUMENT: mở hồ sơ và khoanh đỏ đúng vùng. */
  locator?: SourceLocator;
  /** Có mặt khi sourceType = POLICY, ví dụ "CREDIT-POLICY-3.2". */
  policyClauseId?: string;
  /** Có mặt khi sourceType = RECORD: bảng/bản ghi trong hệ thống. */
  recordRef?: string;
  /** Có mặt khi sourceType = SQL: mô tả truy vấn đã chạy. */
  sqlRef?: string;
  confidence?: number;
}

export interface ReportClaim {
  id: string;
  text: string;
  /** Nguồn của luận điểm này. Rỗng + isInference=false là vi phạm bất biến. */
  citationIds: string[];
  /**
   * Luận điểm SUY LUẬN từ các nguồn khác, không đọc thẳng từ một nguồn.
   * Giao diện đánh dấu riêng để chuyên viên biết đâu là dữ kiện, đâu là suy diễn.
   */
  isInference?: boolean;
  /** Ảnh hưởng tới quyết định: thuận lợi / cảnh báo / trung tính. */
  tone?: 'positive' | 'warning' | 'neutral';
}

export interface DomainReport {
  domain: ReportDomain;
  /** Tên tác nhân đã tạo báo cáo, phục vụ ghi vết. */
  agent: string;
  status: ReportStatus;
  summary: string;
  claims: ReportClaim[];
  citations: ReportCitation[];
  confidence: number;
  /** Lý do khi status khác OK. */
  note?: string;
}

export interface ComposedSection {
  title: string;
  domain?: ReportDomain;
  claims: ReportClaim[];
}

/** Kết luận tổng hợp trình chuyên viên — đã verify chống bịa nguồn. */
export interface ComposedReport {
  question: string;
  /** Đề xuất hành động, viết cho người ra quyết định đọc. */
  recommendation: string;
  decision: 'APPROVE' | 'APPROVE_WITH_CONDITIONS' | 'DECLINE' | 'NEED_MORE_INFO';
  sections: ComposedSection[];
  /** Điều kiện phải hoàn tất trước khi giải ngân. */
  conditions?: string[];
  /** Luận điểm bị loại vì không resolve được nguồn — hiện minh bạch, không giấu. */
  unsupportedClaims?: string[];
  domainStatus: Record<ReportDomain, ReportStatus>;
  warnings?: string[];
}

/** Gói đầy đủ: 3 báo cáo chuyên gia + kết luận tổng hợp. */
export interface ExpertReportBundle {
  composed: ComposedReport;
  domainReports: DomainReport[];
  /**
   * Hồ sơ kèm TOÀN BỘ vùng được các báo cáo trích dẫn.
   *
   * Phải dựng từ chính citation của báo cáo: nếu lấy từ nguồn khác, một trích dẫn
   * có thể trỏ tới vùng không nằm trong danh sách và giao diện mở hồ sơ ra nhưng
   * không khoanh được chỗ nào.
   */
  documents: {
    name: string;
    url: string;
    regions: DocumentRegion[];
  }[];
}

export const DECISION_LABEL: Record<ComposedReport['decision'], string> = {
  APPROVE: 'Đề xuất phê duyệt',
  APPROVE_WITH_CONDITIONS: 'Đề xuất phê duyệt có điều kiện',
  DECLINE: 'Đề xuất từ chối',
  NEED_MORE_INFO: 'Cần bổ sung thông tin',
};

/**
 * Kiểm tra bất biến chống bịa nguồn cho một báo cáo lĩnh vực.
 * Trả về danh sách citationId bị trích nhưng không tồn tại (rỗng = hợp lệ).
 */
export function findDanglingCitations(report: DomainReport): string[] {
  const known = new Set(report.citations.map((citation) => citation.id));
  const dangling: string[] = [];

  for (const claim of report.claims) {
    for (const id of claim.citationIds) {
      if (!known.has(id)) dangling.push(id);
    }
  }
  return dangling;
}
